feat: add `lapack/base/dlapy2`

aayush0325 opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds `lapack/base/dlapy2`

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aayush0325 committed:
feat: add lapack/base/dlapy2

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
feat: add examples tests benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: add docs fodler

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
docs: add README

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: cleanup

aayush0325 committed:
chore: cleanup

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/lib/main.js:
```diff
@@ -0,0 +1,66 @@
+*
+* @example
+* var out = dlapy2( 3.0, 4.0, );
+* // returns 5
```
```suggestion
* // returns 5.0
```

kgryte committed:
docs: fix missing decimal

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/docs/repl.txt:
```diff
@@ -0,0 +1,32 @@
+    overflow.
+
+    If either argument is `NaN` and the other argument is not `+-Infinity`,
+    the function returns `NaN`.
```
I don't believe this is correct, given the implementation.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/README.md:
```diff
@@ -0,0 +1,173 @@
+
+The function has the following parameters:
+
+-   **x**: input number.
+-   **y**: input number.
```
```suggestion
-   **x**: first input number.
-   **y**: second input number.
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/README.md:
```diff
@@ -0,0 +1,173 @@
+
+#### dlapy2( x, y )
+
+Calculate `sqrt( x^2 + y^2 )` in a manner which doesn't cause unnecessary overflow.
```
```suggestion
Computes `sqrt(x^2 + y^2)` in a manner which doesn't cause unnecessary overflow.
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/README.md:
```diff
@@ -0,0 +1,173 @@
+
+# dlapy2
+
+> LAPACK routine to calculate `sqrt( x^2 + y^2 )` in a manner which doesn't cause unnecessary overflow.
```
```suggestion
> LAPACK routine to calculate `sqrt(x^2 + y^2)` in a manner which doesn't cause unnecessary overflow.
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/lib/main.js:
```diff
@@ -0,0 +1,66 @@
+	if ( z === 0.0 ) {
+		return w;
+	}
+
+	return w * ( sqrt( 1.0 + pow( z / w, 2.0 ) ) );
```
```suggestion
	return w * ( sqrt( 1.0 + abs2( z/w ) ) );
```

Use `abs2` rather than `pow` here for squaring a number, as the latter is much more involved and potentially expensive.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/package.json:
```diff
@@ -0,0 +1,69 @@
+{
+  "name": "@stdlib/lapack/base/dlapy2",
+  "version": "0.0.0",
+  "description": "LAPACK routine to calculate sqrt( x^2 + y^2 ) in a manner which doesn't cause unnecessary overflow.",
```
```suggestion
  "description": "LAPACK routine to calculate sqrt(x^2 + y^2) in a manner which doesn't cause unnecessary overflow.",
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/test/test.js:
```diff
@@ -0,0 +1,143 @@
+	t.end();
+});
+
+tape( 'the function returns `NaN` if either argument is `NaN` but not `+-infinity`', function test( t ) {
```
I don't think this test description is correct.

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/test/test.js:
```diff
@@ -0,0 +1,143 @@
+	t.strictEqual( h, PINF, 'returns expected value' );
+
+	h = dlapy2( 1.0e308, 1.0e308 );
+	t.strictEqual( h, 1.4142135623730951e308, 'avoids overflow' );
+
+	t.end();
+});
+
+tape( 'the function avoids underflow', function test( t ) {
+	var h;
+
+	h = sqrt( pow( 1.0e-200, 2 ) + pow( 1e-200, 2 ) );
+	t.strictEqual( h, 0.0, 'returns 0' );
+
+	h = dlapy2( 1.0e-200, 1.0e-200 );
+	t.strictEqual( h, 1.414213562373095e-200, 'avoids underflow' );
```
```suggestion
	t.strictEqual( h, 1.4142135623730951e308, 'returns expected value' );

	t.end();
});

tape( 'the function avoids underflow', function test( t ) {
	var h;

	h = sqrt( pow( 1.0e-200, 2 ) + pow( 1e-200, 2 ) );
	t.strictEqual( h, 0.0, 'returns expected value' );

	h = dlapy2( 1.0e-200, 1.0e-200 );
	t.strictEqual( h, 1.414213562373095e-200, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/lib/index.js:
```diff
@@ -0,0 +1,40 @@
+'use strict';
+
+/**
+* LAPACK routine to calculate sqrt( x^2 + y^2 ) in a manner which doesn't cause unnecessary overflow.
```
```suggestion
* LAPACK routine to calculate `sqrt(x^2 + y^2)` in a manner which doesn't cause unnecessary overflow.
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/lib/index.js:
```diff
@@ -0,0 +1,40 @@
+* var dlapy2 = require( '@stdlib/lapack/base/dlapy2' );
+*
+* var out = dlapy2( 3.0, 4.0 );
+* // returns 5
```
```suggestion
* // returns 5.0
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,45 @@
+// TypeScript Version: 4.1
+
+/**
+* LAPACK routine to calculate sqrt( x^2 + y^2 ) in a manner which doesn't cause unnecessary overflow.
+*
+* @param x - number
+* @param y - number
+* @returns sqrt( x^2 + y^2 )
```
```suggestion
* LAPACK routine to calculate `sqrt(x^2 + y^2)` in a manner which doesn't cause unnecessary overflow.
*
* @param x - number
* @param y - number
* @returns `sqrt(x^2 + y^2)`
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/docs/repl.txt:
```diff
@@ -0,0 +1,32 @@
+
+{{alias}}( x, y )
+    Returns sqrt( x^2 + y^2 ) in a manner which doesn't cause unnecessary
```
```suggestion
    Computes `sqrt(x^2 + y^2)` in a manner which doesn't cause unnecessary
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/docs/repl.txt:
```diff
@@ -0,0 +1,32 @@
+    Returns
+    -------
+    out: number
+        Sqrt( x^2 + y^2 ).
```
```suggestion
        Square of the sum of squares.
```

kgryte reviewed (changes requested):
Left a round of initial comments.

aayush0325 committed:
chore: code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

aayush0325 committed:
chore: code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/test/test.js:
```diff
@@ -0,0 +1,127 @@
+	var h;
+
+	h = sqrt( pow( 1.0e-200, 2 ) + pow( 1e-200, 2 ) );
+	t.strictEqual( h, 0.0, 'returns 0' );
```
```suggestion
	t.strictEqual( h, 0.0, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/lib/main.js:
```diff
@@ -0,0 +1,66 @@
+	if ( z === 0.0 ) {
+		return w;
+	}
+
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/dlapy2/lib/main.js:
```diff
@@ -0,0 +1,65 @@
+/**
+* Returns `sqrt(x^2 + y^2)` in a manner which doesn't cause unnecessary overflow.
+*
+* @private
```
```suggestion
```

kgryte committed:
docs: remove annotation

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (dismissed):
LGTM

kgryte committed:
test: add `NaN` tests

Signed-off-by: Athan <kgryte@gmail.com>
